File: scripts/batch_add_markers.py
#!/usr/bin/env python3
"""
Batch Marker Addition Script for Daemon X Machina: Titanic Scion Interactive Map

This script processes a file with marker data and adds them using add_marker.py.
Each line should be in format: map_id x y category "name" "description"
"""

import subprocess
import sys
import re
from pathlib import Path


def parse_marker_line(line):
    """Parse a line with marker data."""
    line = line.strip()
    if not line or line.startswith('#'):
        return None
    
    # Pattern to match: map_id x y category "name" "description"
    # This handles quoted strings that may contain spaces
    pattern = r'^(\w+)\s+(\d+)\s+(\d+)\s+(\w+)\s+"([^"]*)"(?:\s+"([^"]*)")?'
    match = re.match(pattern, line)
    
    if not match:
        print(f"Warning: Could not parse line: {line}")
        return None
    
    map_id, x, y, category, name, description = match.groups()
    
    return {
        'map_id': map_id,
        'x': int(x),
        'y': int(y),
        'category': category,
        'name': name,
        'description': description or ''
    }


def run_add_marker(marker_data, dry_run=False):
    """Run add_marker.py with the given marker data."""
    cmd = [
        'python', 'scripts/add_marker.py',
        marker_data['map_id'],
        str(marker_data['x']),
        str(marker_data['y']),
        marker_data['category'],
        marker_data['name']
    ]
    
    # Add description if not empty
    if marker_data['description']:
        cmd.append(marker_data['description'])
    
    if dry_run:
        cmd.append('--dry-run')
    
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        return True, result.stdout
    except subprocess.CalledProcessError as e:
        return False, f"Error: {e.stderr}"


def main():
    """Main function."""
    if len(sys.argv) < 2:
        print("Usage: python scripts/batch_add_markers.py <input_file> [--dry-run]")
        sys.exit(1)
    
    input_file = Path(sys.argv[1])
    dry_run = '--dry-run' in sys.argv
    
    if not input_file.exists():
        print(f"Error: File {input_file} does not exist")
        sys.exit(1)
    
    print(f"Processing markers from: {input_file}")
    if dry_run:
        print("DRY RUN mode - no files will be modified")
    print()
    
    # Read and process file
    with open(input_file, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    success_count = 0
    error_count = 0
    
    for line_num, line in enumerate(lines, 1):
        marker_data = parse_marker_line(line)
        if not marker_data:
            continue
        
        print(f"Processing line {line_num}: {marker_data['name']}")
        
        success, output = run_add_marker(marker_data, dry_run)
        if success:
            success_count += 1
            if not dry_run:
                print(f"  ✓ Added successfully")
        else:
            error_count += 1
            print(f"  ✗ Failed: {output}")
        
        # Show minimal output for successful additions unless dry run
        if dry_run or not success:
            print(f"    Output: {output[:200]}...")
        
        print()
    
    print(f"Summary:")
    print(f"  Successful: {success_count}")
    print(f"  Errors: {error_count}")
    print(f"  Total processed: {success_count + error_count}")


if __name__ == '__main__':
    main()

File: scripts/batch_add_markers.js
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { pathToFileURL } from 'node:url';
import { spawnSync } from 'node:child_process';

/**
 * @typedef {Object} MarkerCommand
 * @property {string} mapId
 * @property {string} x
 * @property {string} y
 * @property {string} category
 * @property {string} name
 * @property {string} description
 */

/**
 * @typedef {Object} BatchArguments
 * @property {string} inputFile
 * @property {boolean} dryRun
 */

/**
 * @typedef {Object} AddMarkerResult
 * @property {boolean} success
 * @property {string} output
 */

const ADD_MARKER_SCRIPT = path.resolve('scripts/add_marker.js');

const markerPattern = /^(\w+)\s+(\d+)\s+(\d+)\s+(\w+)\s+"([^"]*)"(?:\s+"([^"]*)")?$/u;

/**
 * @param {string} line
 * @returns {MarkerCommand | null}
 */
function parseMarkerLine(line) {
  const trimmed = line.trim();
  if (trimmed === '' || trimmed.startsWith('#')) {
    return null;
  }

  const match = trimmed.match(markerPattern);
  if (!match) {
    console.warn(`Warning: Could not parse line: ${line.trimEnd()}`);
    return null;
  }

  const [, mapId, x, y, category, name, description] = match;
  return /** @type {MarkerCommand} */ ({
    mapId,
    x,
    y,
    category,
    name,
    description: description ?? '',
  });
}

/**
 * @param {MarkerCommand} markerData
 * @param {boolean} dryRun
 * @returns {AddMarkerResult}
 */
function runAddMarker(markerData, dryRun) {
  const args = [
    ADD_MARKER_SCRIPT,
    markerData.mapId,
    markerData.x,
    markerData.y,
    markerData.category,
    markerData.name,
  ];

  if (markerData.description) {
    args.push(markerData.description);
  }

  if (dryRun) {
    args.push('--dry-run');
  }

  const result = spawnSync('node', args, { encoding: 'utf-8' });

  return {
    success: result.status === 0,
    output: (result.stdout ?? '') + (result.stderr ?? ''),
  };
}

/**
 * @param {string[]} [argv]
 * @returns {BatchArguments}
 */
function parseArguments(argv = process.argv.slice(2)) {
  if (argv.length === 0) {
    throw new Error('Usage: node scripts/batch_add_markers.js <input_file> [--dry-run]');
  }

  const args = [...argv];
  const dryRunIndex = args.indexOf('--dry-run');
  const dryRun = dryRunIndex !== -1;

  if (dryRun) {
    args.splice(dryRunIndex, 1);
  }

  const [inputFile] = args;
  if (!inputFile) {
    throw new Error('Input file path is required');
  }

  return { inputFile: path.resolve(inputFile), dryRun };
}

/**
 * @param {string[]} [argv]
 * @returns {void}
 */
export function main(argv = process.argv.slice(2)) {
  let parsed;
  try {
    parsed = parseArguments(argv);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(message);
    process.exitCode = 1;
    return;
  }

  const { inputFile, dryRun } = parsed;

  if (!fs.existsSync(inputFile)) {
    console.error(`Error: File ${inputFile} does not exist`);
    process.exitCode = 1;
    return;
  }

  console.log(`Processing markers from: ${inputFile}`);
  if (dryRun) {
    console.log('DRY RUN mode - no files will be modified');
  }
  console.log('');

  const lines = fs.readFileSync(inputFile, 'utf-8').split('\n');
  let successCount = 0;
  let errorCount = 0;

  lines.forEach((line, index) => {
    const marker = parseMarkerLine(line);
    if (!marker) {
      return;
    }

    console.log(`Processing line ${index + 1}: ${marker.name}`);
    const { success, output } = runAddMarker(marker, dryRun);

    if (success) {
      successCount += 1;
      if (!dryRun) {
        console.log('  ✓ Added successfully');
      }
    } else {
      errorCount += 1;
      console.log(`  ✗ Failed`);
    }

    if (dryRun || !success) {
      const trimmedOutput = output.trim();
      if (trimmedOutput) {
        const snippet = trimmedOutput.length > 200 ? `${trimmedOutput.slice(0, 200)}...` : trimmedOutput;
        console.log(`    Output: ${snippet}`);
      }
    }
    console.log('');
  });

  const totalProcessed = successCount + errorCount;
  console.log('Summary:');
  console.log(`  Successful: ${successCount}`);
  console.log(`  Errors: ${errorCount}`);
  console.log(`  Total processed: ${totalProcessed}`);
}

/**
 * @param {string | null | undefined} [entry]
 * @param {string} [moduleUrl]
 * @returns {boolean}
 */
function isExecutedDirectly(entry = process.argv[1], moduleUrl = import.meta.url) {
  if (!entry) {
    return false;
  }

  try {
    return pathToFileURL(entry).href === moduleUrl;
  } catch {
    return false;
  }
}

if (isExecutedDirectly()) {
  try {
    main();
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`Error: ${message}`);
    process.exit(1);
  }
}
